src/components/grader/submissions/CodeEditor.js

```diff
@@ -0,0 +1,159 @@
+import "ace-builds/src-noconflict/mode-sql";
+import "ace-builds/src-noconflict/mode-c_cpp";
+import "ace-builds/src-noconflict/mode-java";
+import "ace-builds/src-noconflict/mode-javascript";
+import "ace-builds/src-noconflict/mode-json";
+
+import "ace-builds/src-noconflict/theme-dracula";
+import 'ace-builds/src-min-noconflict/ext-searchbox';
+import apiClient from "../../../api-client";
+
+
+function CodeEditor(props) {
+    const [itemRefs] = useState(() => new MultiRef());
+    const [tabRefs] = useState(() => new MultiRef());
+    const [preVal, setPreVal] = useState("");
+    // eslint-disable-next-line
+    const [theLabels, setTheLabels] = useState([])
+    const [currentTab, setCurrentTab] = useState(0)
+    const [currentTabLabel, setCurrentTabLabel] = useState(props.subList[0])
+    // eslint-disable-next-line
+    const [submissionFiles, setSubmissionFiles] = useState({})
+
+
+    useEffect(() => {
+        if (isEmpty()) {
+            // fetch file here 
+            apiClient.getFile(props.course, props.lab, props.submissionId, currentTabLabel)
+                .then(res => {
+                    setEditor(itemRefs.map.get(currentTab).editor, res.data.file_content);
+                });
+
+        }
+        // eslint-disable-next-line
+    }, [currentTab])
+
+
+    function getLanguageMode(lang) {
+        var ext = lang.substr(lang.lastIndexOf('.') + 1);
+        switch (ext) {
+            case 'py':
+                return 'python';
+            case 'c':
+            case 'cc':
+            case 'cpp':
+                return 'c_cpp';
+            case 'java':
+                return 'java';
+            case 'js':
+                return 'javascript';
+            case 'json':
+                return 'json';
+            case 'sql':
+                return 'sql';
+            default:
+                return 'json';
+        }
+    }
+
+
+    function isEmpty() {
+        return itemRefs.map.get(currentTab).editor.getValue() === ''
+    }
+
+    function setEditor(editor, value) {
+        editor.setValue(value)
+    }
+
+    function handleSelect(index) {
+        setCurrentTab(index)
+        setCurrentTabLabel(theLabels[index])
+    }
+
+
+
+    function getFiles() {
+        const formData = new FormData();
+        for (let key in submissionFiles) {
+            formData.append(key, new Blob([submissionFiles[key]], {
+                type: "text/plain",
+            }))
+        }
+        return formData
+    }
+
+    function saveFiles() {
+        if (submissionFiles) {
+            const formData = getFiles();
+            apiClient.modifySubmissions(props.course, props.lab, props.submissionId, formData)
+                .then(res => {
+                    // console.log(res)
+                    // console.log(submissionFiles)
+                });
+        }
+        return true
+    }
+
+
+    function setTabLabel(theLabel, index) {
+        //handle if exists
+        theLabels[index] = theLabel
```

Comment:
does `theLabels` represent the open source code files?

Reply:
I don't get the question. Yet, `theLabels` are  the labels for `Tabs` based on their indices.